server/src/main/java/org/elasticsearch/index/IndexMode.java

```diff
@@ -40,6 +51,45 @@ void validateWithOtherSettings(Map<Setting<?>, Object> settings) {
             }
         }
 
+        @Override
+        public void validateWithSource(DocumentParserContext context) {
+            validateTimestamp(context);
+        }
+
+        private void validateTimestamp(DocumentParserContext context) {
+            IndexableField[] fields = context.rootDoc().getFields(TIMESTAMP_FIELD);
```

Comment:
I wonder if we'd be better off validating this in `DateFieldMapper` somehow. We can get the single-valued-ness validation with `addWithKey`. We could, maybe, have `IndexMode` get a method like:
```
public abstract LongConsumer dateValidator(String date, Resolution resolution);
```

In non-tsdb it'd just `d -> {}` and in tsdb it'd check.

I think it'd be safe to read the bounds from the IndexSettings with the same volatile stuff you have now. In an ideal world we'd update the mapping if the bounds changed and we wouldn't need these extra volatile reads. But I think that's a bigger change.

Reply:
@imotov do you think it makes sense to check after the fact like we do here or to push something to the mapper?

Reply:
As I mentioned before I would like this to be consistent with the rest of data streams. Since in data streams this is configurable, we need to make it configurable, unless we will decide otherwise in our meeting on Thursday.

Reply:
I prefer DataStreamTimestampFieldMapper to be a common TimestampFieldMapper, then time_series can reuse it. After you discuss with folks on data streams and come to a conclusion, then we can discuss how to modify it.